<div style="border:solid green 2px; padding: 20px">
    
<div class="alert alert-success">
    
<b>Review summary, v.2</b> 
    
Hello Byron.
    
Thanks for the update. Now everything looks good, you've fixed all the critical comments. Your project is complete, congratulations.
    
Thanks again for your work on it. And good luck to you with your next projects and further studies.

<div style="border:solid blue 2px; padding: 20px">
  
**Hello Byron.**

My name is Yulia, and I will be reviewing your project. 

You will find my comments in coloured cells marked as 'Reviewer's comment'. The cell colour will vary based on the contents - I am explaining it further below. 

**Note:** Please do not remove or change my comments - they will help me in my future reviews and will make the process smoother for both of us. 

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment</b> 
    
Such comment will mark efficient solutions and good ideas that can be used in other projects.
</div>

<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment</b> 
    
The parts marked with yellow comments indicate that there is room for optimisation. Though the correction is not necessary it is good if you implement it.
</div>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment</b> 
    
If you see such a comment, it means that there is a problem that needs to be fixed. Please note that I won't be able to accept your project until the issue is resolved.
</div>

You are also very welcome to leave your comments / describe the corrections you've done / ask me questions, marking them with a different colour. You can use the example below: 

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>

<div style="border:solid green 2px; padding: 20px">
    
<div class="alert alert-success">
<b>Review summary</b> 
    
Byron, thanks for submitting the project. You've done a very good start.
    
- I can see that you already know many python methods that allow to thoroughly study dataframes 
- You know how to find and remove missing values and duplicates
- You can write your own functions
    
<div class="alert alert-danger">

However, there are a few critical comments that need to be corrected. You will find them in the red-coloured cells in relevant sections.
    
Could you please fix them and send the project back to me?


</div>
    
You may also work on the yellow comments. If you have any questions please write them when you return your project. 
    
I'll be looking forward to getting your updated notebook.

## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Title and introduction are essential parts of the project. Make sure you do not forget to include it in your further projects. 
    
It is optimal if introduction consists of:
    
- brief description of the situation;
- goal of the project;
- description of the data we are going to use.
</div>


## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [1]:
# importing pandas
import pandas as pd

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    

It's a good practice to upload all necessary libraries in the first code cell, separately from datasets (so that you do not have to reload the dataset if you need to add another library). 

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [2]:
# reading the files and storing them to df
df = pd.read_csv('/datasets/movies_and_shows.csv')

Print the first 10 table rows:

In [3]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [4]:
# obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Well done, you are using good methods to get the first impression of the data.

The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Please note that it is highly recommended to add a conclusion / summary after each section and describe briefly your observations and / or major outcomes of the analysis.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [5]:
# the list of column names in the df table
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [6]:
# renaming columns 
df = df.rename({'   name': 'name','release Year': 'release_year', 'imdb v0tes': 'imdb_votes', 'imdb sc0re': 'imdb_score', 'Character': 'character', 'r0le': 'role', 'TITLE': 'title','  Type': 'type'}, axis=1)


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
This is a good way to rename the columns. And you've set proper names, in line with Python naming convention.

Check the result. Print the names of the columns once more:

In [7]:
# checking result: the list of column names
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [8]:
# calculating missing values
df.isna().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [9]:
# dropping rows where columns with title, scores and votes have missing values
# df = df.dropna()
cleaned_df = df.copy()
cleaned_df.dropna(inplace=True)
df.dropna(inplace=True)

<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment, v. 2</b> 

You did not really need the first two rows in the code above. You are cleaning the missing values and saving the result to `df` dataframe (which you are using further) in the third line of the code.
    

    
<a class="tocSkip"></a><s>
<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 

You need to save the result. Now, you are just displaying the dataframe without the missing values. Please correct.

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [10]:
# counting missing values
df.isna().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [11]:
# counting duplicate rows
df.duplicated().sum()

6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [12]:
# Produce table with duplicates (with original rows included) and review last 5 rows
duplicates_rows = df[df.duplicated()]
duplicates_rows.tail()


name                 character   role    title   type  \
85569       Jessica Cediel           Liliana Navarro  ACTOR  Lokillo  MOVIE   
85570  Javier Gardeaz?­bal  Agust??n "Peluca" Ort??z  ACTOR  Lokillo  MOVIE   
85571        Carla Giraldo            Valery Reinoso  ACTOR  Lokillo  MOVIE   
85572  Ana Mar??a S?­nchez                   Lourdes  ACTOR  Lokillo  MOVIE   
85577         Isabel Gaona                    Cacica  ACTOR  Lokillo  MOVIE   

       release_year      genres  imdb_score  imdb_votes  
85569          2021  ['comedy']         3.8        68.0  
85570          2021  ['comedy']         3.8        68.0  
85571          2021  ['comedy']         3.8        68.0  
85572          2021  ['comedy']         3.8        68.0  
85577          2021  ['comedy']         3.8        68.0

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [13]:
# removing duplicate rows
df.drop_duplicates(inplace = True)


Check for duplicate rows once more to make sure you have removed all of them:

In [14]:
# checking for duplicates
df.duplicated().sum()

0

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Great, you found and removed the duplicates.

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [15]:
# viewing unique type names
df['type'].sort_values().unique()

array(['MOVIE', 'SHOW', 'movies', 'shows', 'the movie', 'tv', 'tv series',
       'tv show', 'tv shows'], dtype=object)

Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [16]:
# function for replacing implicit duplicates    
def replace_wrong_show(wrong_shows_list, correct_show):
    for wrong_show in wrong_shows_list:
        df['type'].replace({wrong_show: correct_show},inplace = True)

Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [17]:
# removing implicit duplicates
replace_wrong_show(wrong_shows_list=['shows', 'SHOW', 'tv show','tv shows', 'tv series', 'tv'], correct_show = 'show')

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [18]:
# viewing unique genre names
df['type'].unique()

array(['MOVIE', 'the movie', 'show', 'movies'], dtype=object)

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
You function worked properly - all the implicit duplicates for 'show' are gone.

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [19]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name
shows_after_1999 = df[df['release_year'] >= 1999]

<a class="tocSkip"></a><s>
<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment, v. 1</b> 

You've filtered the data for the required time period correctly. Although, the title of your new dataframe is not optimal - please note the double underscore `__` between `released` and `after`.

Notes from tutor:
You should keep in mind which dataframe contain what value
df = pd.read_csv() # original data
df_copy = df.copy()
Step 1: data clean & preprocess  -> df_copy
Step 2: apply filters -> 
year_movie_1999_df = df_copy[filter1]
different_tile_df = df_copy[filter2]

In [20]:
# repeat conditional indexing so df has only shows (movies are removed as result)
shows_after_1999 = shows_after_1999[shows_after_1999['type'] == 'show'] # save the result
shows_after_1999.head() # for making sure the result is correct

name                character   role      title  type  \
1664       Jeff Probst           Himself - Host  ACTOR   Survivor  show   
2076     Mayumi Tanaka  Monkey D. Luffy (voice)  ACTOR  One Piece  show   
2077      Kazuya Nakai     Roronoa Zoro (voice)  ACTOR  One Piece  show   
2078     Akemi Okamura             Nami (voice)  ACTOR  One Piece  show   
2079  Kappei Yamaguchi            Usopp (voice)  ACTOR  One Piece  show   

      release_year                                             genres  \
1664          2000                                        ['reality']   
2076          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2077          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2078          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2079          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   

      imdb_score  imdb_votes  
1664         7.4     24687.0  
2076         8.8    117129.0  
2077         8.8    117129.0  
2078         8.8    117129.0  
2079         8.8    117129.0

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
✔️ 
    
<a class="tocSkip"></a><s>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 

There are several moments that need to be corrected here:
    
1. You need to save the result of your conditional filtering.
2. You need to filter the dataframe that you created at a previous step, not `df`.
3. You need to filter for 'show', not 'shows'.

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [21]:
# rounding column with scores
shows_after_1999.loc[:,'imdb_score'] = shows_after_1999.loc[:,'imdb_score'].round()
#checking the outcome with tail()
shows_after_1999.loc[:,'imdb_score'].tail()

85433    8.0
85434    8.0
85435    8.0
85436    8.0
85437    8.0
Name: imdb_score, dtype: float64

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
✔️ 
    
<a class="tocSkip"></a><s>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 

Here and below you will need to work with the filtered dataframe.

It is now time to identify outliers based on the number of votes.

In [22]:
# Use groupby() for scores and count all unique values in each group, print the result
shows_after_1999.groupby('imdb_score').nunique()

name  character  role  title  type  release_year  genres  \
imdb_score                                                             
2.0           24         18     2      5     1             4       5   
3.0           27         27     2      4     1             3       4   
4.0          174        108     2     30     1            13      19   
5.0          581        419     2     94     1            12      68   
6.0         2365       1896     2    347     1            19     212   
7.0         4342       3813     2    593     1            24     304   
8.0         4194       4028     2    539     1            23     329   
9.0          542        498     2     65     1            20      50   
10.0           8          8     1      1     1             1       1   

            imdb_votes  
imdb_score              
2.0                  5  
3.0                  4  
4.0                 30  
5.0                 91  
6.0                329  
7.0                568  
8.0                527  
9.0                 65  
10.0                 1

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [23]:
# filter dataframe using two conditions (scores to be in the range 4-9)
shows_averages = shows_after_1999[ (shows_after_1999['imdb_score'] >=4) & (shows_after_1999['imdb_score'] <=9)]
# group scores and corresponding average number of votes, reset index and print the result
shows_averages = shows_averages.groupby('imdb_score')['imdb_votes'].agg('mean').reset_index()

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
✔️ All good - you've left only the rows with imdb score between 4 and 9 and grouped the data correctly.
    
<a class="tocSkip"></a><s>



Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [24]:
shows_averages

imdb_score     imdb_votes
0         4.0    5277.583333
1         5.0    3143.942568
2         6.0    3481.717322
3         7.0    8727.068211
4         8.0   30299.460967
5         9.0  126904.109515

In [25]:
# round column with averages
shows_averages= shows_averages.round()
# rename columns
shows_averages = shows_averages.rename(columns ={'imdb_score': 'average_score', 'imdb_votes': 'average_votes'})
shows_averages

average_score  average_votes
0            4.0         5278.0
1            5.0         3144.0
2            6.0         3482.0
3            7.0         8727.0
4            8.0        30299.0
5            9.0       126904.0

In [26]:
#print df in descending order
shows_averages.sort_values(by=['average_score', 'average_votes'],ascending= False)

average_score  average_votes
5            9.0       126904.0
4            8.0        30299.0
3            7.0         8727.0
2            6.0         3482.0
1            5.0         3144.0
0            4.0         5278.0

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
✔️ 
    
<a class="tocSkip"></a><s>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 

You are trying to round the column called `avg_vote`, but you don't have such a column.

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.

You can input the conclusion here.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Overall conclusion is an important part, where we should include the summary of the outcomes of the project.
    
Please note that you will need to write one for every project you make.